src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -0,0 +1,721 @@
+        IonTypeID valueTid = typeIds[typeIdByte];
+        if (!valueTid.isValid) {
+            throw new IonException("Invalid type ID.");
+        } else if (valueTid.type == IonTypeID.ION_TYPE_ANNOTATION_WRAPPER) {
+            if (isAnnotated) {
+                throw new IonException("Nested annotation wrappers are invalid.");
+            }
+            if (minorVersion == 0 ? readAnnotationWrapperHeader_1_0(valueTid) : readAnnotationWrapperHeader_1_1(valueTid)) {
+                return true;
+            }
+            hasAnnotations = true;
+            return readHeader(buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK, true, valueMarker);
+        } else {
+            long endIndex = minorVersion == 0
+                ? calculateEndIndex_1_0(valueTid, isAnnotated)
+                : calculateEndIndex_1_1(valueTid, isAnnotated);
+            if (isAnnotated) {
+                validateAnnotationWrapperEndIndex(endIndex);
+            }
+            setMarker(endIndex, markerToSet);
+            if (endIndex > limit) {
+                event = Event.NEEDS_DATA;
+                return true;
+            }
+        }
+        markerToSet.typeId = valueTid;
+        if (event == Event.START_CONTAINER) {
+            if (minorVersion == 0) {
+                prohibitEmptyOrderedStruct_1_0(markerToSet);
+            }
+            return true;
+        }
+        return event == Event.START_SCALAR;
+    }
+
+    /**
+     * Doubles the size of the cursor's container stack.
+     */
+    private void growContainerStack() {
+        Marker[] newStack = new Marker[containerStack.length * 2];
+        System.arraycopy(containerStack, 0, newStack, 0, containerStack.length);
+        for (int i = containerStack.length; i < newStack.length; i++) {
+            newStack[i] = new Marker(-1 ,-1);
+        }
+        containerStack = newStack;
+    }
+
+    /**
+     * Push a Marker representing the current container onto the stack.
+     */
+    private void pushContainer() {
+        containerIndex++;
+        if (containerIndex >= containerStack.length) {
+            growContainerStack();
+        }
+        parent = containerStack[containerIndex];
+    }
+
+    @Override
+    public Event stepIntoContainer() {
+        if (valueTid == null || valueTid.type.ordinal() < LIST_TYPE_ORDINAL) {
+            throw new IonException("Must be positioned on a container to step in.");
+        }
+        // Push the remaining length onto the stack, seek past the container's header, and increase the depth.
+        pushContainer();
+        parent.typeId = valueTid;
+        parent.endIndex = valueTid.isDelimited ? -1 : valueMarker.endIndex;
+        valueTid = null;
+        event = Event.NEEDS_INSTRUCTION;
+        reset();
+        return event;
+    }
+
+    @Override
+    public Event stepOutOfContainer() {
+        if (parent == null) {
+            // Note: this is IllegalStateException for consistency with the other binary IonReader implementation.
```

Comment:
Does that mean that you'd rather have a different behavior?

Reply:
Yes, I'd prefer that we always throw some kind of `IonException` from all these methods, reserving `IllegalStateException` for states that cannot occur unless we, the library authors, broke something. Although, I guess this usage does technically align with the official documentation on `IllegalStateException`. Also, this brings to my attention that stepIn() when not on a container also throws `IllegalStateException`, so I will make that change for consistency too.

Reply:
`IllegalStateException` is definitely an exception to indicate a programmer error and reserving it to mean a programmer error within the library (as opposed to by a user) makes sense. Maybe just indicate the preference here in the comment.

Reply:
Addressed in https://github.com/amazon-ion/ion-java/pull/538/files